HttpListenerFactory: improve port selection

tmds opened the pull request:
This speeds up test execution of System.Net.HttpListener.Tests by about 15 seconds.

Pre:
```
System.Net.HttpListener.Tests  Total: 638, Errors: 0, Failed: 0, Skipped: 16, Time: 141.388s
```
Post:
```
System.Net.HttpListener.Tests  Total: 638, Errors: 0, Failed: 0, Skipped: 16, Time: 127.527s
```

stephentoub commented on src/System.Net.HttpListener/tests/HttpListenerFactory.cs:
```diff
@@ -7,15 +7,34 @@
+                if (limits.Length == 2)
+                {
+                    int.TryParse(limits[0], out _minPort);
+                    int.TryParse(limits[1], out _maxPort);
```
If these TryParses fail, won't _minPort and/or _maxPort be set to 0?  Seems like you'd want these to be:
```C#
int tmp;
if (int.TryParse(limits[0], out tmp) _minPort = tmp;
if (int.TryParse(limits[1], out tmp) _maxPort = tmp;
```

stephentoub commented on src/System.Net.HttpListener/tests/HttpListenerFactory.cs:
```diff
@@ -7,15 +7,34 @@
+        const string ProcPortRangeFileName = "/proc/sys/net/ipv4/ip_local_port_range";
+        private static int _minPort;
+        private static int _maxPort;
+        private static readonly Random _random = new Random();
```
Nit: s_ prefix on statics.  And the s_minPort and s_maxPort could be readonly as well, right?

stephentoub commented on src/System.Net.HttpListener/tests/HttpListenerFactory.cs:
```diff
@@ -7,15 +7,34 @@
-        private static readonly object s_createListenerLock = new object();
-        private static int s_nextPort = MinPort;
+        const int MaxStartAttempts = 50;
+        const string ProcPortRangeFileName = "/proc/sys/net/ipv4/ip_local_port_range";
```
Nit: private on both consts

stephentoub commented on src/System.Net.HttpListener/tests/HttpListenerFactory.cs:
```diff
@@ -7,15 +7,34 @@
-        const int MinPort = 1025;
-        private static readonly object s_createListenerLock = new object();
-        private static int s_nextPort = MinPort;
+        const int MaxStartAttempts = 50;
```
Where'd the value of 50 come from?

stephentoub commented on src/System.Net.HttpListener/tests/HttpListenerFactory.cs:
```diff
@@ -33,55 +52,50 @@ internal HttpListenerFactory(string hostname = "localhost", string path = null)
-            lock (s_createListenerLock)
+            for (int attempt = 0; attempt < MaxStartAttempts; attempt++)
             {
-                foreach (int port in Enumerable.Range(s_nextPort, IPEndPoint.MaxPort - s_nextPort + 1).Concat(Enumerable.Range(MinPort, s_nextPort - MinPort)))
```
Why the switch from this to the random approach?

tmds commented:
@stephentoub I had not seen your review comments and already changed the "user range" instead of the "dynamic range" since that was in use before and I also dropped the code that figures out where the dynamic range starts to simplify the PR. I've addressed the comments that still apply.

The main source of contention was that except for the windows managed case, every factory would start trying ports from the same minimum. Incrementing in all cases would have solved that already.
Random has the (minor) advantage, that it doesn't compete with other processes which are applying the incremental strategy too.
The value of 50 was a guess. I thought it was a lot, but some tests were already failing with "Could not reserve a port for HttpListener" so I've increased it to 500.

tmds commented:
The NETFX build has failed with "Could not reserve a port for HttpListener" for every commit, even with max attempts at 500. It seems to have a strong preference for the incremental strategy. I wonder why.

tmds commented:
```
12:42:59         System.Exception : Could not reserve a port for HttpListener
12:42:59         ---- System.Net.HttpListenerException : Access is denied
```
Probably the port range is restricted.
Should I revert back to the incremental strategy? Or what range could be used on Windows?

krwq commented on src/System.Net.HttpListener/tests/HttpListenerFactory.cs:
```diff
@@ -7,15 +7,24 @@
+        private const int MaxStartAttempts = 500;
+        private static int s_minPort;
+        private static int s_maxPort;
+        private static readonly Random s_random = new Random();
```
random is not thread safe

krwq replied:
didn't notice these are tests - I do not know if there are any negative implications of that then

krwq commented on src/System.Net.HttpListener/tests/HttpListenerFactory.cs:
```diff
@@ -33,55 +42,50 @@ internal HttpListenerFactory(string hostname = "localhost", string path = null)
+                    {
+                        // If this is not an HttpListenerException or SocketException, something very wrong has happened, and there's no point
+                        // in trying again.
+                        break;
```
should we simply:
```csharp
else
{
    throw;
}
```

?

Do we expect any other exceptions here?

stephentoub commented:
> Should I revert back to the incremental strategy?

I'd just stick with that.

tmds commented:
@stephentoub I played with this a bit and settled on the current implementation. Can you do another review?

stephentoub commented:
@dotnet-bot test Windows x86 Release Build please
